[LoMaply] iP

LoMaply opened the pull request:
# 🪶 Jyuuni 🪶 
> "Easy, Fast, Efficient." -Me
> "Truly one of the chatbots of all time." -Myself
> "For free?" -I
## 💬 Description
Jyuuni is a helpful chatbot aiming to help users organise their tasks. Some of its more useful features include:
- **Todo/Deadline/Event** 3 preformatted task types users can add to their list.
- **Mark/Unmark** tasks as completed or uncompleted.
- **Search** for specific tasks using keywords.

***And more*** all in an easy to use command line interface.

## ▶️ Running the project
1. [Download/Fork](https://github.com/nus-cs2103-AY2324S1/ip) the project.
2. Open it in an IDE of choice.
3. Navigate to **src/main/java/gui/Launcher.java**.
4. Run the following main method.
```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```
### 📔 Additional Note
To change the default location of the data file used to load and store chatbot data, navigate to 
**src/main/java/duke/Duke.java** and change the file location used in the storage field of the Duke constructor.
```java
public Duke() {
    storage = new Storage("./data");
    taskList = new TaskList();
}
```

## ⌛ ~~Future~~ Tentative Features & Updates
- [x] Add/Delete/Mark/Unmark tasks.
- [x] List/Search tasks.
- [x] Load/Save data across sessions.
- [x] An all new GUI for easier user access.
- [ ] View schedule for a chosen day???


Could [this](https://www.youtube.com/watch?v=E2xNlzsnPCQ&t=111s) be butterfly?

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

LoMaply committed:
Add greeting

LoMaply committed:
Add end command, chatbot repeats user input

LoMaply committed:
Add list command, other user inputs are added to task list

LoMaply committed:
Add mark/unmark commands, Task class

LoMaply committed:
Add todo, deadline, event classes for task types

LoMaply committed:
Add automated UI test

LoMaply committed:
Add error handling for wrong user input

LoMaply committed:
Add delete command, modify task array to use ArrayList

LoMaply committed:
Add minor style changes

LoMaply committed:
Add Storage class to read/write task data into file

LoMaply committed:
Add error checking for loadData in Storage class, chatbot prompt for file error

LoMaply committed:
Merge branch 'branch-Level-7'

LoMaply committed:
Add support for LocalDateTime in Deadline/Event types, Update test cases

LoMaply committed:
Merge branch 'branch-level-8'

LoMaply committed:
Refactor code for additional OOP

LoMaply committed:
Merge branch 'branch-A-MoreOOP'

LoMaply committed:
Add class packages

LoMaply committed:
Merge remote-tracking branch 'origin/add-gradle-support'

LoMaply committed:
Add gradle support

LoMaply committed:
Add JUnit tests

LoMaply committed:
Add additional Javadocs

LoMaply committed:
Add additional code formatting

LoMaply committed:
Add find command to Parser class, add test cases

LoMaply committed:
Merge branch 'branch-A-JavaDoc'

LoMaply committed:
Merge branch 'branch-A-CodingStandard'

LoMaply committed:
Resolve merge conflict

LoMaply committed:
Add gradle checkstyle and fix all style issues

LoMaply committed:
Merge tag 'A-CheckStyle'

LoMaply committed:
Add enum for Ui and fix some javadoc comments

LoMaply committed:
Fix additional javadocs comments

LoMaply committed:
Add GUI display with JavaFx and code formerly used for CLI display

andytoh1 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+    public static boolean parseCommand(String userInput, TaskList taskList) {
+        String[] splitInput = userInput.split(" ", 2); // Limit 2 to only separate out command word.
+        switch (splitInput[0]) {
+        case "mark":
```
I like that the coding standard of not indenting case clauses is adhered to.

LoMaply committed:
Modify storage to return string for CLI output, add JUnit test

andytoh1 commented on src/main/java/taskutil/Task.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return Formatted string, to be further modified by overriding methods in child classes.
+     */
+    public String toFileString() {
```
Perhaps naming these methods "getString" and "getFileString" would better adhere to the requirement of method names being verbs.

LoMaply replied:
Personally I feel that using **getString** and **getFileString** might suggest that the Task object has a String variable that can be "obtained" (which to be fair it has a title variable but it isn't meant to be accessed directly).

Thus using **toString** and **toFileString** keeps it more in line with Java's default **toString** method for objects where we are simply converting an object to its string representation.

But thanks for the suggestion anyway!

andytoh1 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            return false;
+        default:
+            Ui.output("Sorry, I don't recognise this comment :(");
+            break;
```
Perhaps this break statement can be omitted, as was done in the Storage class.

andytoh1 reviewed (commented):
Very clean code with stringent and graceful error handling.

LoMaply committed:
Add minor GUI changes

LoMaply committed:
Merge branch 'branch-Level-10'

ruiyangzh commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,19 @@
+    public void parseDate_exceptionThrown_success() {
+        try {
+            Parser.parseDate("idk");
+        } catch (DateTimeParseException e) {
```
To make this test more complete you can have a fail() after line 14, to represent a test fail if the code executes without the correct exception. You can also use assertThrows instead of a try/catch.

ruiyangzh commented on src/test/java/taskutil/TaskListTest.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Test
+    public void changeStatus_listToStringData_correctOutput() {
```
Nice compliant method names on the Junit tests.

ruiyangzh commented on src/main/java/taskutil/Task.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @param isCompleted New status of task.
+     */
+    public void changeStatus(boolean isCompleted) {
```
maybe a bit of a nitpick, but should it be setStatus? (since it could be the same as the current status, e.g)

ruiyangzh commented on src/main/java/taskutil/Task.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    public String getStatus() {
+        return (isDone ? "X" : " ");
+    }
```
getStatus seems to imply the method is returning a boolean, not a String.

ruiyangzh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,137 @@
+            // Using 1 split statement that splits by /(from|to) allows wrong combos like "/to x /from x".
+            try {
+                String[] eventVar = splitInput[1].split(" /from ", 2);
+                String[] times = eventVar[1].split(" /to ", 2);
```
Another nitpick, should this be in plurals (eventVars)?

ruiyangzh reviewed (commented):
Generally not too much to say, well-thought out. more junit test types can be used to test more efficiently and extensively.

LoMaply committed:
Add assertion tests to several methods in different classes

Some methods have invariants that are not explicitly checked for
as they are assumed to be fulfilled when the methods are called

Assertion tests help ensure these invariants are fulfilled without
having to type out JUnit test cases, or exception handling

LoMaply committed:
Refactor code to be more readable

Several new methods are extracted from long methods such as
parseCommand from the parser class to help with readability.

parseCommand itself is modified along with several other methods
to return a String directly rather than rely on the Ui class.

This keeps the Ui class methods restricted to only functions
that involve adding and deleting tasks which simplifies the code.

LoMaply committed:
Merge pull request #1 from LoMaply/branch-CodeQuality

Refactor code to be more readable

LoMaply committed:
Merge branch 'master' into branch-A-Assertions

LoMaply committed:
Minor edit

LoMaply committed:
Merge pull request #2 from LoMaply/branch-A-Assertions

Add assertion tests to several methods in different classes

LoMaply committed:
Add Continuous Integration with GitHub Actions

LoMaply committed:
Add Schedule command, change Task class to be abstract

LoMaply committed:
Modify GUI, Add help command, Change chatbot lines

LoMaply committed:
Add more error handling, load data GUI text

LoMaply committed:
Small GUI change

LoMaply committed:
Edit README, find command is no longer case-sensitive

LoMaply committed:
Edit README

LoMaply committed:
Minor changes for checkstyle

LoMaply committed:
Minor change to StorageTest
